037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,177 @@
+
+A new `spec.kafka.autoRebalance` field could be added for this purpose.
+Because we are talking about auto-rebalancing, the corresponding field could also be not limited to the rebalancing on scaling but a more general “continuous” auto-rebalancing to enable on the cluster. 
+To specify when running the auto-rebalancing, a `spec.kafka.autoRebalance.mode` field could be used by listing if running auto-rebalancing only on scaling, during the normal execution or both.
+Having “continuous” auto-rebalancing is out of scope for this proposal, the field is added for allowing a future implementation.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      mode: [running, scaling] # or just one of the two
+```
+
+Finally, the user should be able to define what are the [goals and the other rebalancing related configurations](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaRebalance-reference) which are normally described via a `KafkaRebalance` custom resource.
+In any case, for rebalancing the cluster, an actual `KafkaRebalance` custom resource has to be automatically created by the cluster operator.
+
+### New `KafkaRebalanceTemplate` custom resource
+
+A new `KafkaRebalanceTemplate` custom resource could be a solution allowing the user to define the rebalancing configuration he wants but leaving to the operator the creation of the corresponding `KafkaRebalance` custom resource.
+This new CRD should be quite similar to the `KafkaRebalance`.
+The cluster operator has to copy over the configuration from the template to the actual rebalancing resource.
+Compared to the rejected alternative, it’s about exporting the rebalancing configuration out of the Kafka custom resource avoiding to pollute that resource and increase the size.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalanceTemplate
+metadata:
+  name: my-rebalance-template
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  # NOTE: mode and brokers fields are not available 
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The user has the flexibility to specify goals as well as all the other options like he already does today on a `KafkaRebalance` custom resource.
+The cluster operator will use that configuration for the `KafkaRebalance` but adding the `spec.mode` and `spec.brokers` accordingly.
+
+It would also contain a `spec.autoApproval` field, as described by this [proposal](https://github.com/strimzi/proposals/pull/56), to specify if auto-approval is desired or not when auto-rebalancing is enabled.
```

Comment:
So really, when you say "auto-rebalancing" you are actually saying "auto-proposal-generation + auto-approval"? If a user doesn't set `autoapproval` to true but has auto-rebalacing enabled then the KafkaRebalance CRs will be created and sit there waiting for approval?

I wonder if we should call that behaviour out explicitly by not calling the behaviour in this propsoal "auto-rebalancing" and instead calling it `autoProposalGeneration` or something similar? This would better match today's UX where everything is approved by the user. 

Making this proposal about auto-proposal-generation would be cleaner in my opinion. This would also fit better with the auto-approval proposal. They can be implemented separately but when both are implemented we will then have "auto-rebalancing".

Reply:
The manual approval was removed from the proposal, so now it's an "auto-proposal-generation" + "auto-approval" as you said and "auto-rebalancing" sounds to be fine now.